src/ClpIrV1Decoder.cpp

```diff
@@ -0,0 +1,224 @@
+
+auto ClpIrV1Decoder::decode(size_t begin_idx, size_t end_idx) -> emscripten::val {
+    if (m_log_events.size() < end_idx || begin_idx >= end_idx) {
+        return emscripten::val::null();
+    }
+
+    std::string message;
+    constexpr size_t cDefaultReservedMessageLength{512};
+    message.reserve(cDefaultReservedMessageLength);
+    emscripten::val const results{emscripten::val::array()};
+    std::span<clp::ir::LogEvent<clp::ir::four_byte_encoded_variable_t> const> log_events_span(
+            m_log_events.begin() + static_cast<std::ptrdiff_t>(begin_idx),
+            m_log_events.begin() + static_cast<std::ptrdiff_t>(end_idx)
+    );
+    for (auto const& log_event : log_events_span) {
+        message.clear();
+
+        // TODO: Replace below handlers code by an OSS decoding method once it's added in the future
+        auto const constant_handler = [&](std::string const& value, size_t begin_pos, size_t length
+                                      ) { message.append(value, begin_pos, length); };
+
+        auto const encoded_int_handler = [&](clp::ir::four_byte_encoded_variable_t value) {
+            message.append(clp::ffi::decode_integer_var(value));
+        };
+
+        auto const encoded_float_handler
+                = [&](clp::ir::four_byte_encoded_variable_t encoded_float) {
+                      message.append(clp::ffi::decode_float_var(encoded_float));
+                  };
+
+        auto const dict_var_handler
+                = [&](std::string const& dict_var) { message.append(dict_var); };
+
+        try {
+            clp::ffi::ir_stream::generic_decode_message<true>(
+                    log_event.get_logtype(),
+                    log_event.get_encoded_vars(),
+                    log_event.get_dict_vars(),
+                    constant_handler,
+                    encoded_int_handler,
+                    encoded_float_handler,
+                    dict_var_handler
+            );
+        } catch (clp::ffi::ir_stream::DecodingException const& e) {
+            SPDLOG_ERROR("Failed to decode. Error Code: {}", e.get_error_code());
+            break;
+        }
+
+        constexpr size_t cLogLevelNone{0};
+        constexpr size_t cLogLevelBeginIdx{cLogLevelNone + 1};
+        constexpr size_t cLogLevelPositionInMessages{1};
+        size_t log_level{cLogLevelNone};
+        for (size_t i = cLogLevelBeginIdx; i < cLogLevelNames.size(); ++i) {
+            if (message.substr(cLogLevelPositionInMessages).starts_with(cLogLevelNames[i])) {
+                log_level = i;
+                break;
+            }
+        }
+
+        m_ts_pattern.insert_formatted_timestamp(log_event.get_timestamp(), message);
+
+        EM_ASM(
+                { Emval.toValue($0).push([UTF8ToString($1), $2, $3, $4]); },
+                results.as_handle(),
+                message.c_str(),
+                log_event.get_timestamp(),
+                log_level,
+                begin_idx + (&log_event - log_events_span.data()) + 1
```

Comment:
I assume this is the log index? We should just add a counter to keep track of the indices... Try to avoid ptr arithmetic whenever it's possible

Reply:
Right. That was the reason we started with the index access approach rather than the span approach in the first place. 

Still, `span` + a counter doesn't sounds too bad. Let me add it.